### DISCUSSION TITLE
missing berry - wire - read without register

### DISCUSSION THREAD

**Discussion by NiklausS** [arendst/Tasmota#20297] [Q&A]
Ts'me again,
thanks for you help so far. 
I've written now my new driver for SHT3X and found that the wire implementation is missing a 'read without register'.
I've checked out all the wire functions but there seems to be no `wire.readbytes( <addr>, <number of bytes> )`
Fortunately the SHT3X is flexible enough to ignore the register.
![image](https://github.com/arendst/Tasmota/assets/15640381/4708fde9-5458-44ea-99a5-24372a8ca08a)

in the picture the leading Write,Addr + Reg 00  is not needed and the Read,Add ... is fully sufficient. I do remember other i2c devices to work the same.
Don't know if this is worth an issue tough.
Nik

**Comment by barbudor** [comment]
Does the address disturb the SHT3x ?

If you are looking for the equivalent of Arduino's `Wire.requestFrom(address, quantity)`, it is available as the internal Berry `wire._request_from(...)`
Not tested so not guaranteed but the Berry code for `wire.readbytes()` is here:
https://github.com/arendst/Tasmota/blob/b4f1aafc4cc10b8f037cdfefe47c2cea4711f70d/lib/libesp32/berry_tasmota/src/embedded/Wire.be#L9-L19
So you have to combine yourself calls to `wire._request_from()` and a loop around `wire._available()` and `wire._read()`
Juts like the Tasmota native driver: 
https://github.com/arendst/Tasmota/blob/15e181ac53f285da66b26389d593871c9966784d/tasmota/tasmota_xsns_sensor/xsns_14_sht3x.ino#L100-L103


May I ask why you are bothering writing a Berry driver while there is a native Tasmota driver for SHT3x ?

**Reply by NiklausS** [reply]
i did try the` _request_from() `but the `_read()` does not return a value but only 1 (for one bit read ?) 
and when looking at the logicanalyzer the packet was not what i expected. I have tried some combinations but did not succeed with a pleasant solution. The Low-level  Functions seem to a bit strange.

**About the drivers in Berry** - i've started compiling linux kernels in the early 90s - and this went on until micropython in 2023 with the esp32; you're never in line with the current. The sexy thing on Tasmota and Berry is : install tasmota.bin (-lvgl) with the browser * and add a suitable set of berry scrtips or pages.jsonl and and your done. **0 dependency on the PC**. And it works great, plus after 3 years i will need only the tasmota.bin and a couple of berryscripts to tune something - and they will be still on the device. 

*though i still use esptool :-) 

hers's the latest driver for SHT3X - don't know where's the drawback to the 'native' driver, bit it's a much more flexible for me.

```berry
sht3x.be

#-
 - I2C driver written in Berry
 -
 - Support for SHT3X device (based on TMP117)
 - 20. Dez 2023
 -#

import string

class SHT3X: Driver
  static name        = "SHT3X"       # sensor name
  static sensor_addr = 0x44           # I2C bus address for sensor

  var wire            #- if wire == nil then the module is not initialized -#
  var humidity        # temp result in °C
  var temp            # temp result in °C
  var tempU           # temp unit °F / °C
  var tempR           # Temp display resolution

  def init()
    self.wire = tasmota.wire_scan(self.sensor_addr)
    if self.wire
      print("I2C: "..self.name.." detected on bus "+str(self.wire.bus))
    else
      # todo: if we don't have a device do we need to return false? 
      #       and if so do i need the !self.wire check in all the fuctions?
      return
    end
    #- write config Register / initialize Sensor -#
    self.tempR = int(tasmota.cmd("TempRes").find("TempRes"))
    tasmota.add_rule("TempRes", /val -> self.set_Res(val) )
  end

  # set the temp resolution - todo: don't know how to put it at lambda directly
  def set_Res(v) self.tempR = int(v) end

  #- get's the temp and humidity from the device -#
  def read_temp()
    self.wire.write(self.sensor_addr, 0x2c, 0x06, 1)
    # todo: the 13ms depend on the accuracy add logic to separate request and result
    tasmota.delay(13)

    var b = self.wire.read_bytes(self.sensor_addr, 0x00, 6)
    # todo: b=bytes() - the read is not successfull - retry or disable driver

    self.humidity = 100* b.get(3,-2)/65535
    self.temp = 175.0* b.get(0,-2)/65535 -45

    if tasmota.get_option(8) == 1
      self.temp = self.temp * 1.8 + 32
      self.tempU="°F"
    else
      self.tempU="°C"
    end

    tasmota.publish_rule(string.format(
      '{"%s":{"Temperature":%f,"Humidity":%i}}', self.name,self.temp, self.humidity ))
  end

  #- trigger a read every second called by tasmota-driver -#
  def every_second() self.read_temp() end

  #- display sensor value in the web UI  called by tasmota-driver -#
  def web_sensor()
    tasmota.web_send_decimal(string.format(
      "{s}%s Temperatur {m}%."..self.tempR.."f %s{e}"..
      "{s}%s Humidity {m}%i %%{e}", self.name, self.temp, self.tempU, self.name, self.humidity))
  end

  #- add sensor value to teleperiod  called by tasmota-driver -#
  def json_append()
    tasmota.response_append(string.format(
      ",\"%s\":{\"Temperatur\":%i,\"Humidity\":%i} ", self.name, self.temp, self.humidity))
  end

end
sht3x = SHT3X()
tasmota.add_driver(sht3x)


```

**Reply by sfromis** [reply]
The drawback is mainly doing work without getting more from it. But if that's your way of having fun....

You are also losing out on Tasmota features for selecting number of decimals for temperature and humidity, instead of hardcoding for zero decimals. For better integration, preferably use same names, like `Temperature` instead of `Temp`; then you also get the temperature unit automatically part of the JSON payload for sensors.

And you **cannot** use `tasmota.bin` on ESP32, that would be `tasmota32.bin` which already has support for SHT sensors in the precompiled variant.

**Reply by NiklausS** [reply]
ok, updated the code with your suggestions, and cleaned up some more, thx for pointing this out.

I am actually using the **13.3.0(lvgl-haspmota)** with a M5stack Basic, a M5Stick (gray) and a TTGO oled - so there is no support for SHT ootb. The SHT driver is just a simple blueprint. I'm working on a readout of 3 - 6 INA226 with different shunt resistors, eventually INA 3221, and here it gets tricky to cover all the specific settings with a stock driver. Development and rollout of these drives is so much easier without recompiling and uploading - in the simplest case i can modify a driver just by copy, rename and little codechange.
And upon the next tasmota release, with some simple sanity checks i can deploy **the solid stock release/tasmota32-lvg.bin**.
I'm also happy to contribute my work if there is a suitable place.
 
with the **UrlFetch https://...** it's so nice to deploy small missing pieces. It could also be an idea to have all the drivers deployed this way, in the Tasmota gui just picking up the required drivers and your done. Not to mention the easiness of testing.

However - i appreciate your work very much. Thanks

**Reply by sfromis** [reply]
If you want "maximally simple" deploy of drivers you like to spend time creating, you could package them as a [Tasmota Application](https://tasmota.github.io/docs/Tasmota-Application/). The `.tapp` files are automatically loaded after a restart, and you need no `autoexec.be` editing. You could also push such files to a Tasmota device instead of doing the fetch from inside Tasmota.

Still, it is quite a bit more work than simply building Tasmota with a mix of already available features. Not at all hard to do.

**Comment by s-hadinger** [comment]
Back to the original question (sorry I'm late to the party). You can't just call `_request_from()` out of the blue. You must always send a write/command to the device first. Otherwise they don't know which byte in the sequence to send back. A write/command will reset the byte counter to zero on the device, and it will then send the full sequence.
